### PR TITLE
Changed Elements Sample

### PR DESCRIPTION
**Pull request by Ellord207**
Sample demonstrates the Changed Elements API.

### CODE REVIEW

File: src/frontend-samples/version-compare-sample/VersionCompareWidget.tsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import "common/samples-common.scss";
+import * as React from "react";
+import { NamedVersion, VersionCompareApi } from "./VersionCompareApi";
+import "./VersionCompare.scss";
+
+export const VersionCompareWidget: React.FunctionComponent = () => {
+  // State Declarations
+  const iModelConnection = useActiveIModelConnection();
+  const [isRequest, setIsRequest] = React.useState<boolean>(false);
+  const [selectVersion, setVersion] = React.useState<NamedVersion>(VersionCompareApi.namedVersions[1]);
```
Comment (diegopinate) [inline src/frontend-samples/version-compare-sample/VersionCompareWidget.tsx:19]: Nitpick: can the sample change iModels? If so, is it guaranteed namedVersions array's length will always be >= 2?

File: src/frontend-samples/version-compare-sample/VersionCompareApi.ts [hunk 2]
```diff
@@ -0,0 +1,159 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+import { DbOpcode, Id64Array, Id64String } from "@bentley/bentleyjs-core";
+import { ColorDef, FeatureAppearance } from "@bentley/imodeljs-common";
+import { AuthorizedFrontendRequestContext, EmphasizeElements, FeatureOverrideProvider, FeatureSymbology, IModelApp, NotifyMessageDetails, OutputMessagePriority, Viewport } from "@bentley/imodeljs-frontend";
+import { VersionCompareClient } from "./VersionCompareClient";
+
+export interface NamedVersion {
```
Comment (diegopinate) [inline src/frontend-samples/version-compare-sample/VersionCompareApi.ts:10]: Any reason to not use [Version](https://www.itwinjs.org/reference/imodelhub-client/imodelhubclient/version/)?

File: src/frontend-samples/version-compare-sample/VersionCompareClient.ts [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+      }
+  * `````
+  */
+  public static async getNamedVersions(top = 100): Promise<any | undefined> {
+    const { requestContext, iModelId } = VersionCompareClient._projectContext;
+
+    const accessToken = await VersionCompareClient.getAccessToken();
+    if (accessToken === undefined)
+      return undefined;
+    const url = ` https://api.bentley.com/imodels/${iModelId}/namedversions?$top=${top}`;
```
Comment (diegopinate) [inline src/frontend-samples/version-compare-sample/VersionCompareClient.ts:137]: Any reason to not use the clients in iModelJs? Something like:
```
const namedVersions: Version[] = await IModelApp.iModelClient.versions.get(
  requestContext,
  iModelId
);
 ```

File: src/frontend-samples/version-compare-sample/VersionCompareWidget.tsx [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+export class VersionCompareWidgetProvider implements UiItemsProvider {
+  public readonly id: string = "VersionCompareWidgetProvider";
+
+  public provideWidgets(_stageId: string, _stageUsage: string, location: StagePanelLocation, _section?: StagePanelSection): ReadonlyArray<AbstractWidgetProps> {
+    const widgets: AbstractWidgetProps[] = [];
+    if (location === StagePanelLocation.Right) {
+      widgets.push(
+        {
+          id: "VersionCompareWidget",
+          label: "Version Compare Controls",
```
Comment (diegopinate) [inline src/frontend-samples/version-compare-sample/VersionCompareWidget.tsx:62]: Same comment about naming `Version Compare` -> `Changed Elements`?

File: src/frontend-samples/version-compare-sample/readme.md [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+- Provide [code examples](./VersionCompareWebApi.ts) for calling the Version Compare APIs.
+- Colorize elements based on last changes made to it.
+
+## Description
+
+The Changeset Ids are needed as arguments to request the Version Comparison.  We get an iModel's Named Versions by requesting them using a [Named Versions API](https://developer.bentley.com/api-groups/data-management/apis/imodels/operations/get-imodel-named-versions/) call.  This response does contain the Changeset Id, but it is contained in the href to the Changeset API and must be parsed out.
+
+While the API allows to compare any two named versions, this sample allows only comparisons against the latest version of the iModel.
+
+The Changed Elements API returns a list of changed elements and the type of operations that updated them.  This information is passed to a [FeatureOverrideProvider](https://www.itwinjs.org/reference/imodeljs-frontend/views/featureoverrideprovider/) that colorizes the elements based on the operations.  For another example using feature overrides, see the [Emphasize Elements](../emphasize-elements-sample/readme.md) sample.
```
Comment (diegopinate) [inline src/frontend-samples/version-compare-sample/readme.md:19]: Can you add a link to [ChangedElements](https://www.itwinjs.org/reference/imodeljs-common/entities/changedelements/)? This is the JSON format returned by the API.

File: src/frontend-samples/version-compare-sample/VersionCompareClient.ts [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        "newChecksums": [
+          [362149254]
+        ],
+        "parentIds": ["0"],
+        "parentClassIds": ["0"]
+      }
+    }
+   * ```
+   */
+  public static async getChangedElements(startChangesetId: string, endChangesetId: string): Promise<any | undefined> {
```
Comment (diegopinate) [inline src/frontend-samples/version-compare-sample/VersionCompareClient.ts:66]: We may want to cast the result to [ChangedElements](https://www.itwinjs.org/reference/imodeljs-common/entities/changedelements/)

File: src/frontend-samples/version-compare-sample/VersionCompareApi.ts (hunk 2 above)
Comment (Ellord207) [reply src/frontend-samples/version-compare-sample/VersionCompareApi.ts:10]: I was unaware it existed.  Thank you for pointing this out.

Comment (Ellord207) [reply src/frontend-samples/version-compare-sample/VersionCompareApi.ts:10]: After looking at the type declaration, it requires wsgId, and ecId.  It feels strange to use a type that forces me to assign useless data when the property names in the response don't match those in the type.

File: src/frontend-samples/version-compare-sample/VersionCompareClient.ts (hunk 3 above)
Comment (Ellord207) [reply src/frontend-samples/version-compare-sample/VersionCompareClient.ts:137]: Your last to comment only makes sense in context of this.  Was unaware this was possible.  Will investigate adoption.

File: src/frontend-samples/version-compare-sample/VersionCompareClient.ts (hunk 6 above)
Comment (Ellord207) [reply src/frontend-samples/version-compare-sample/VersionCompareClient.ts]: Casted results in the compareChangesets method in ChangedElementApi.ts

File: src/frontend-samples/version-compare-sample/readme.md (hunk 5 above)
Comment (Ellord207) [reply src/frontend-samples/version-compare-sample/readme.md]: Was unaware of the typing.  Now using it in compareChangesets method in ChangedElementApi.ts

File: src/frontend-samples/version-compare-sample/VersionCompareWidget.tsx (hunk 4 above)
Comment (Ellord207) [reply src/frontend-samples/version-compare-sample/VersionCompareWidget.tsx]: Renamed whole sample to Changed Elements

File: src/frontend-samples/version-compare-sample/VersionCompareWidget.tsx (hunk 1 above)
Comment (Ellord207) [reply src/frontend-samples/version-compare-sample/VersionCompareWidget.tsx]: Good catch.  Validating Named Version length now.

File: src/frontend-samples/version-compare-sample/VersionCompareApi.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    }
+    const insertFeature = FeatureAppearance.fromRgb(ColorDef.green);
+    const updateFeature = FeatureAppearance.fromRgb(ColorDef.blue);
+    this._insertOp.forEach((id) => overrides.overrideElement(id, insertFeature));
+    this._updateOp.forEach((id) => overrides.overrideElement(id, updateFeature));
+    overrides.setDefaultOverrides(ComparisonProvider._defaultAppearance);
+  }
+}
+
+export class VersionCompareApi {
```
[the remark this answers is not in the document]
Comment (Ellord207) [reply src/frontend-samples/version-compare-sample/VersionCompareApi.ts]: Renamed whole sample to Changed Elements.

File: src/frontend-samples/version-compare-sample/readme.md [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+## Purpose
+
+- Demonstrate requesting Named Versions and their changeset Ids.
+- Provide [code examples](./VersionCompareWebApi.ts) for calling the Version Compare APIs.
+- Colorize elements based on last changes made to it.
+
+## Description
+
+The Changeset Ids are needed as arguments to request the Version Comparison.  We get an iModel's Named Versions by requesting them using a [Named Versions API](https://developer.bentley.com/api-groups/data-management/apis/imodels/operations/get-imodel-named-versions/) call.  This response does contain the Changeset Id, but it is contained in the href to the Changeset API and must be parsed out.
```
[the remark this answers is not in the document]
Comment (Ellord207) [reply src/frontend-samples/version-compare-sample/readme.md]: Renamed whole sample to Changed Elements.

File: src/frontend-samples/version-compare-sample/VersionCompareApi.ts (hunk 2 above)
Comment (Ellord207) [reply src/frontend-samples/version-compare-sample/VersionCompareApi.ts]: After adopting the IModelClient.verison.get method to populate the Named Version, using this type makes sense, and I'm now doing so.

File: src/frontend-samples/version-compare-sample/VersionCompareClient.ts (hunk 3 above)
Comment (Ellord207) [reply src/frontend-samples/version-compare-sample/VersionCompareClient.ts]: This is much simpler than previous method.  Thank you for the suggestion.

**Comment by williamkbentley** [comment]
Your clients use `any` a bunch. I'm not very fond of `any`. Is that because the clients were automatically generated? Is it possible to use proper types?

**Comment by Ellord207** [comment]
> Your clients use `any` a bunch. I'm not very fond of `any`. Is that because the clients were automatically generated? Is it possible to use proper types?

I was following the format of Clash review, but have no issue adding stronger types.

**Review by williamkbentley — APPROVED** [review]

**Pull request merged**